src/plugins/controls/public/control_group/editor/control_editor.tsx

```diff
@@ -55,25 +60,75 @@ interface EditControlProps {
 }
 
 export const ControlEditor = ({
-  onTypeEditorChange,
-  removeControl,
-  updateTitle,
-  updateWidth,
   embeddable,
   isCreate,
-  onCancel,
-  factory,
-  onSave,
   title,
   width,
+  onSave,
+  onCancel,
+  removeControl,
+  updateTitle,
+  updateWidth,
+  onTypeEditorChange,
 }: EditControlProps) => {
+  const { controls } = pluginServices.getServices();
+  const { getControlTypes, getControlFactory } = controls;
+
+  const [selectedType, setSelectedType] = useState(
+    !isCreate && embeddable ? embeddable.type : getControlTypes()[0]
+  );
+  const [defaultTitle, setDefaultTitle] = useState<string>();
   const [currentTitle, setCurrentTitle] = useState(title);
   const [currentWidth, setCurrentWidth] = useState(width);
-
   const [controlEditorValid, setControlEditorValid] = useState(false);
-  const [defaultTitle, setDefaultTitle] = useState<string>();
 
-  const ControlTypeEditor = factory.controlEditorComponent;
+  const getControlTypeEditor = (type: string) => {
+    const factory = getControlFactory(type);
+    const ControlTypeEditor = (factory as IEditableControlFactory).controlEditorComponent;
+    return ControlTypeEditor ? (
+      <ControlTypeEditor
+        onChange={onTypeEditorChange}
+        setValidState={setControlEditorValid}
+        initialInput={embeddable?.getInput()}
+        setDefaultTitle={(newDefaultTitle) => {
+          if (!currentTitle || currentTitle === defaultTitle) {
+            setCurrentTitle(newDefaultTitle);
+            updateTitle(newDefaultTitle);
+          }
+          setDefaultTitle(newDefaultTitle);
+        }}
+      />
+    ) : null;
+  };
+
+  const getTypeButtons = (controlTypes: string[]) => {
+    return controlTypes.map((type) => {
+      const factory = getControlFactory(type);
+      const icon = (factory as EmbeddableFactoryDefinition).getIconType?.();
+      const tooltip = (factory as EmbeddableFactoryDefinition).getDescription?.();
+      const menuPadItem = (
+        <EuiKeyPadMenuItem
+          id={`createControlButton_${type}`}
+          data-test-subj={`create-${type}-control`}
+          label={(factory as EmbeddableFactoryDefinition).getDisplayName()}
+          isSelected={selectedType === type}
+          onClick={() => {
+            setSelectedType(type);
+          }}
+        >
+          <EuiIcon type={!icon || icon === 'empty' ? 'controlsHorizontal' : icon} size="l" />
+        </EuiKeyPadMenuItem>
+      );
+
+      return tooltip ? (
+        <EuiToolTip content={tooltip} position="bottom">
+          {menuPadItem}
+        </EuiToolTip>
+      ) : (
+        menuPadItem
+      );
+    });
```

Comment:
This creates the menu for choosing control types:

<img src="https://user-images.githubusercontent.com/8698078/158694315-6b08aba1-dc07-4205-b21c-0c15910f798e.gif" width="500px"/>

Note the following:
- If `getIcon` is not defined in the control type's factory, the icon will default to `controlsHorizontal`
- If `getDescription` is not defined in the control type's factory, the tooltip will not be shown.